### ISSUE TITLE
Integrate with an event-driven architecture

### ISSUE THREAD

**Issue by jhagege** [dagster-io/dagster#2415]
Hi, is there any possibility to integrate a dagster pipeline with an event-driven architecture. 
I'd like to have steps trigger external jobs on different infra and block until those jobs communicate their response through existing pubsub infrastructure. 
On successful receival of message, I'd like to unblock the step and pass him the result from the pubsub message.
Can you think of any possibility to handle it ? 
Thanks.

**Comment by mgasner** [comment]
Hi @jhagege, you can do this right now if you're willing to have your blocking steps sleep while waiting to hear back from the jobs they triggered. A more sophisticated approach with support in the execution engines for deferring step execution until some external event occurs would require some work in core to make possible -- but you can see kind of how that might go if you look at the  `RetryRequested` object. cc @alangenfeld

**Comment by jhagege** [comment]
Thanks @mgasner , I'm wondering if I could pass an asyncio.Future() instance to a Dagster solid and wait for its result to be ready (it would be updated from an event handler outside the pipeline) ?
How to pass such a future to the pipeline so that it is accessible both from the solid and from outside ?

**Comment by mgasner** [comment]
huh, interesting idea. i think that you could do something like this with a resource that could communicate with the external infra and return futures. so a solid would receive some kind of serialized representation of the event it was waiting on, call `context.resources.external_infra.get_future(serialized_event_repr)` or equiv, and then wait on the future it received

**Comment by mgasner** [comment]
i'm not certain that the solids play nicely with asyncio

**Comment by catherinewu** [comment]
closing; duplicate of https://github.com/dagster-io/dagster/issues/2268

another relevant feature for event-driven architectures is sensors: https://docs.dagster.io/overview/schedules-sensors/sensors

**Comment by amardeep** [comment]
Reading sensor docs, it seems for eg. s3 storage sensors work by polling in a tight loop run by default every 30s.

Instead, it would be nice if it could integrate with services like google pub sub which can send notifications. For eg.,
here is the link for pubsub notifications for google cloud storage - https://cloud.google.com/storage/docs/pubsub-notifications

I am not sure if this use case is covered in #2268 or is already supported by existing sensor apis.

**Comment by sryza** [comment]
Thanks for the feedback @amardeep.  I'm not super familiar with Google Pub/Sub, but I just took a gander at their docs.  If I understand correctly, subscribers interact with it by polling - e.g. https://cloud.google.com/pubsub/docs/reference/rest/v1/projects.subscriptions/pull.  So it seems possible to write a sensor that would poll it.  Or is there a push-based API you had in mind that I'm missing?

**Comment by amardeep** [comment]
Thanks @sryza. Looks like polling makes most sense. Pubsub has push api, but that requires running a [https endpoint](https://cloud.google.com/pubsub/docs/push) - probably not useful for cases like storage sensors.

**Comment by benedikt-buchert** [comment]
I think you could also run a cloud function that triggers your pipelines: https://docs.dagster.io/_apidocs/libraries/dagster-graphql#dagster_graphql.DagsterGraphQLClient.submit_job_execution
